### PR TITLE
[Refactor:CourseMaterials] Rewrite Routes to Use IDs

### PR DESCRIPTION
**Pull request by cjreed121**
### Please check if the PR fulfills these requirements:

* [x] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant

### What is the current behavior?
Right now the site uses an encoded path which is placed in the URL to view and download course materials. This adds extra encoding to the system and makes long URLs.

### What is the new behavior?
Instead of specifying a path you can specify a course material ID in the route which will reduce URL lengths and remove encoding.

### CODE REVIEW

**Review by HenrikLam — CHANGES_REQUESTED** [review]
Looks good with a couple minor suggestions.

File: site/app/controllers/course/CourseMaterialsController.php [hunk 1]
```diff
@@ -145,31 +193,21 @@ public function modifyCourseMaterialsFileTimeStamp($filenames, $newdatatime): Js
...
         $success = false;
 
-        foreach ($data as $filename) {
-            if (!isset($filename)) {
-                $this->core->redirect($this->core->buildCourseUrl(['course_materials']));
-            }
+        $courseMaterial = $this->core->getCourseEntityManager()->getRepository(CourseMaterial::class)
+            ->findOneBy(['id' => $id]);
+        $courseMaterials = $this->core->getCourseEntityManager()->getRepository(CourseMaterial::class)
+            ->findAll();
```
[the remark this answers is not in the document]
Comment (cjreed121) [reply site/app/controllers/course/CourseMaterialsController.php:202]: Before I continue on with the rest of the function I immediately check if they are null and if they are not then I will call a recursive function with them. If I set them in the recursive function I may be doing extra database work when I don't need to.

[the remark this answers is not in the document]
Comment (HenrikLam) [reply site/app/controllers/course/CourseMaterialsController.php:204]: That does bring up the point that maybe we will want that `courseMaterial` entity in the future. I will approve this PR then.

**Review by HenrikLam — APPROVED** [review]
Looks good to me!

**Pull request merged**